src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -0,0 +1,972 @@
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "No recipients");
+    }
+    
+    fromtaddr_ = CBitcoinAddress(fromAddress);
+    isfromtaddr_ = fromtaddr_.IsValid();
+    isfromzaddr_ = false;
+
+    libzcash::PaymentAddress addr;
+    if (!isfromtaddr_) {
+        CZCPaymentAddress address(fromAddress);
+        try {
+            PaymentAddress addr = address.Get();
+
+            // We don't need to lock on the wallet as spending key related methods are thread-safe
+            SpendingKey key;
+            if (!pwalletMain->GetSpendingKey(addr, key)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid from address, no spending key found for zaddr");
+            }
+            
+            isfromzaddr_ = true;
+            frompaymentaddress_ = addr;
+            spendingkey_ = key;
+        } catch (std::runtime_error e) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, string("runtime error: ") + e.what());
+        }
+    }
+}
+
+AsyncRPCOperation_sendmany::~AsyncRPCOperation_sendmany() {
+}
+
+void AsyncRPCOperation_sendmany::main() {
+    if (isCancelled())
+        return;
+
+    set_state(OperationStatus::EXECUTING);
+    start_execution_clock();
+
+    bool success = false;
+
+    try {
+        success = main_impl();
+    } catch (Object objError) {
+        int code = find_value(objError, "code").get_int();
+        std::string message = find_value(objError, "message").get_str();
+        set_error_code(code);
+        set_error_message(message);
+    } catch (runtime_error e) {
+        set_error_code(-1);
+        set_error_message("runtime error: " + string(e.what()));
+    } catch (logic_error e) {
+        set_error_code(-1);
+        set_error_message("logic error: " + string(e.what()));
+    } catch (...) {
+        set_error_code(-2);
+        set_error_message("unknown error");
+    }
+
+    stop_execution_clock();
+
+    if (success) {
+        set_state(OperationStatus::SUCCESS);
+    } else {
+        set_state(OperationStatus::FAILED);
+    }
+
+    std::string s = strprintf("async rpc %s finished (status=%s", getId(), getStateAsString());
+    if (success) {
+        s += strprintf(", tx=%s)\n", tx_.ToString());
+    } else {
+        s += strprintf(", error=%s)\n", getErrorMessage());
+    }
+    LogPrintf("%s",s);
+}
+
+// Notes:
+// 1. Currently there is no limit set on the number of joinsplits, so size of tx could be invalid.
+// 2. Note selection is not optimal
+// 3. Spendable notes are not locked, so an operation running in parallel could also try to use them
+bool AsyncRPCOperation_sendmany::main_impl() {
+
+    bool isSingleZaddrOutput = (t_outputs_.size()==0 && z_outputs_.size()==1);
+    bool isPureTaddrOnlyTx = (isfromtaddr_ && z_outputs_.size() == 0);
+    CAmount minersFee = ASYNC_RPC_OPERATION_DEFAULT_MINERS_FEE;
+
+    // Regardless of the from address, add all taddr outputs to the raw transaction.
+    if (isfromtaddr_ && !find_utxos(isSingleZaddrOutput)) {
```

Comment:
Added clarification and better error message reporting in 0fce02d0b8f092770300754068273704107d7c85.  Currently, sending from coinbase utxos to a single zaddr is allowed, and any change goes to the same zaddr.  If sending to multiple zaddrs, where does the change goto?  We can't have a transparent output as change and the API does not allow us to specify a zaddr for change.